### PR TITLE
Activity log: Fix null activityTargetTs schema

### PR DESCRIPTION
**Pull request by sirreal**
Three atomic changes, two janitorial commits labeled as `[OPTIONAL]`

* Allow null-type values for activityTargetTs
* [OPTIONAL] Use explicit null default
* [OPTIONAL] Use simple list of types status schema

> Because of an indentation change if may be helpful to view the diff for this PR [with whitespace changes ignored](https://github.com/Automattic/wp-calypso/pull/23672/files?w=1) (`?w=1)`

## Issue

Some `activityTargetTs` were defaulting to `undefined` causing schema invalidations:

schema-invalidation

## Testing
1. Tests pass
1. In dev environment, ensure activity data is loaded and persisted
1. Reload and ensure that no schema invalidation warnings appear in the console

### CODE REVIEW

**Comment by sirreal** [comment]
It's worth noting that the transformer in `fromApi` could omit `activityTargetTs` rather than using `undefined`/`null` as the value. That would be a valid approach requiring no persistence schema changes.

**Comment by dmsnell** [comment]
@sirreal I think I like the approach of only conditionally loading it inside of `fromApi` - less data, less memory

would you want me to update this branch for that of would you like to? do you disagree?

**Comment by dmsnell** [comment]
Conditionally set in e51d48d

@sirreal do you mind going over this to confirm it's what you meant? To me I'm a bit curious now if the API response could also be `null` and that ends up throwing off the state schema because `hasOwnProperty()` would still hold. If that's the case I'll update it to also check against `null`

Thanks!

File: client/state/activity-log/log/schema.js [hunk 1]
```diff
@@ -32,10 +32,8 @@ const activityItemsSchema = {
 				activityIsDiscarded: { type: 'boolean' },
 				activityIsRewindable: { type: 'boolean' },
 				activityName: { type: 'string' },
-				activityStatus: {
-					oneOf: [ { type: 'string' }, { type: 'null' } ],
-				},
-				activityTargetTs: { type: 'number' },
+				activityStatus: { type: [ 'null', 'string' ] },
+				activityTargetTs: { type: [ 'null', 'number' ] },
```
Comment (sirreal) [inline client/state/activity-log/log/schema.js]: `'null'` can be dropped here now.

File: client/state/activity-log/log/schema.js [hunk 2]
```diff
@@ -32,10 +32,8 @@ const activityItemsSchema = {
 				activityIsDiscarded: { type: 'boolean' },
 				activityIsRewindable: { type: 'boolean' },
 				activityName: { type: 'string' },
-				activityStatus: {
-					oneOf: [ { type: 'string' }, { type: 'null' } ],
-				},
-				activityTargetTs: { type: 'number' },
+				activityStatus: { type: [ 'null', 'string' ] },
```
Comment (sirreal) [inline client/state/activity-log/log/schema.js]: @dmsnell Worth doing something similar with `activityStatus` and `rewindId`? Based on the schema they're in the same boat 🙂

File: client/state/data-layer/wpcom/sites/activity/from-api.js [hunk 3]
```diff
@@ -43,33 +43,36 @@ export function transformer( apiResponse ) {
...
+			activityIsDiscarded: item.is_discarded,
+			activityIsRewindable: item.is_rewindable,
+			rewindId: item.rewind_id,
+			activityName: item.name,
+			activityStatus: item.status,
+			activityTitle: item.summary,
+			activityTs: Date.parse( published ),
+			activityDescription: parseBlock( item.content ),
+		},
+		object.hasOwnProperty( 'activityTargetTs' ) && {
```
Comment (sirreal) [inline client/state/data-layer/wpcom/sites/activity/from-api.js]: I think this should be `target_ts`? I'll get that in another commit.

**Comment by sirreal** [comment]
👍 Thanks, I like this direction 🙂 

> To me I'm a bit curious now if the API response could also be null and that ends up throwing off the state schema because hasOwnProperty() would still hold. If that's the case I'll update it to also check against null

This seems to be OK for the `object.target_ts`, however the `item.status` was returning `null` from the API.

The schema is leaner and cleaner, as is the data. I don't see and schema invalidation on my test site's data now and all the properties appear to be correctly populated.

@dmsnell Do you mind giving it a final look on [calypso.live](https://calypso.live/?branch=fix/activity-log-targetts-schema) before I merge to make sure there aren't any inadvertent issues I may have missed?

**Comment by sirreal** [comment]
> I don't see and schema invalidation on my test site's data now and all the properties appear to be correctly populated.

I spoke too soon. Seeing issues with `rewindId` now 😞 

Hold off and I'll fix…

**Comment by sirreal** [comment]
> Hold off and I'll fix…

Appears to have been some stale state, I think this is ready after all.

File: client/state/data-layer/wpcom/sites/activity/from-api.js [hunk 4]
```diff
@@ -43,33 +43,35 @@ export function transformer( apiResponse ) {
...
+			activityIcon: get( item, 'gridicon', DEFAULT_GRIDICON ),
+			activityId: item.activity_id,
+			activityIsDiscarded: item.is_discarded,
+			activityIsRewindable: item.is_rewindable,
+			activityName: item.name,
+			activityTitle: item.summary,
+			activityTs: Date.parse( published ),
+			activityDescription: parseBlock( item.content ),
+		},
+		item.hasOwnProperty( 'rewind_id' ) && { rewindId: item.rewind_id },
```
Comment (dmsnell) [inline client/state/data-layer/wpcom/sites/activity/from-api.js]: `rewind_id` won't be falsey if it's legitimate so we can shorten here…

```js
item.rewind_id && { rewindId: item.rewind_id },
```

Comment (sirreal) [reply client/state/data-layer/wpcom/sites/activity/from-api.js]: 👍 947dd903f1b4a26532d2673729ed1e67660977f4

**Comment by dmsnell** [comment]
I didn't find any errors and I tried persisting state from many months in the past. Code looks good too (after many fruitful iterations!).

:shipit:

**Pull request merged**